### DISCUSSION TITLE
Making a start on the theming wiki :)

### DISCUSSION THREAD

**Discussion by AzumaHazuki** [lxqt/lxqt#2359] [Ideas]
I've begun writing up a theming wiki. It's...not very good yet, and I could use any help people can give, but it's got most of the basics so far. Visible <a href=https://github.com/lxqt/lxqt/wiki/Creating-LXQt-Themes>here</a>, and I'm sorry about how much space it takes up in the sidebar @_@

Can anyone expand on this, and maybe explain what the <a href=https://github.com/lxqt/lxqt/discussions/2354#discussioncomment-4543656>set of cryptic selectors</a> seen here does? If that's a magic bullet for consistency in theming, we should definitely have a subsection in the wiki for it.

**Comment by tsujan** [comment]
Very nice! Keep up the good work.

**Comment by AzumaHazuki** [comment]
Thank you, but I can't do this alone. Leaving aside that no one seems to know all the widgets and toplevels, there are still some matters of questionable accuracy in there. In particular:

- What files are mandatory, besides lxqt-panel.qss? *Is* that one even mandatory?

- If there are no (other) mandatory files, how does one prevent, for example, the line  `QMenu { some_stuff: something; }` in one file from clobbering a `QMenu { other_stuff: something_else; }` stanza in another file?

- What _does_ that <a href=https://github.com/lxqt/lxqt/discussions/2354#discussioncomment-4543656>big ugly block of selectors</a> do, exactly? What widgets at what levels of inheritance does it affect?

- What is the difference between the selectors `LXQtPanel Plugin`, `LXQtPanelPlugin`, and just plain `Plugin` and where do we use each?

And so on. I'm kind of surprised these things weren't written down somewhere, and no, I don't believe in "self-documenting code," at least not to this extent.

**Reply by tsujan** [reply]
Questions can be asked here. Sooner or later, they'll be answered. As for the above questions:

 * `lxqt-panel.qss` should exist. If the other files don't exist, the active Qt widget style will be used for their corresponding apps (`lxqt-runner`, `lxqt-notificationd` and `lxqt-config`).
 * Files don't know about each other.
 * `QWidget > QWidget` means "any direct child widget of a widget", `QWidget > QToolButton` means "any tool-button that is a direct child of a widget", and so on.
 * `LXQtPanel` is for the panel itself, while `Plugin` means "any plugin". I don't know why some themes have `LXQtPanelPlugin` — maybe it's a remnant of old times — but it seems redundant. ~We don't have `LXQtPanel Plugin`.~

EDIT: Sorry, `LXQtPanel Plugin` is what its name says; it's the same as `Plugin` in `lxqt-panel.qss`.

> I'm kind of surprised these things weren't written down somewhere

The lack of enough manpower is the cause, as was mentioned elsewhere. That's why we need (and very much appreciate) contributions for these things.

**Comment by isf63** [comment]
I ran `rg "setObjectName" plugin-*` on the lxqt-panel repository and it returned these identifiers:
[output.txt](https://github.com/lxqt/lxqt/files/10326432/output.txt)

**Reply by AzumaHazuki** [reply]
@isf63 Awesome! That's really helpful and dug up a whole bunch of object identifiers I had no idea even existed. 

So, if I understand this right, anything passed through the function `setObjectName(QStringLiteral("AThingThatExists"));` will produce an addressable widget with the identifier `#AThingThatExists` and it will be themeable according to the base Qt widget class that it was derived from?

@tsujan, so does this mean LXQt specifically looks for files named `lxqt-{panel,notificationd,runner,[configuration?]}.qss` and will _not_ pick up on `someotherfile.qss?` And that configuring a widget unique to, let's say, the LXQt Runner like `#commandEd` in `lxqt-panel.qss` rather than `lxqt-runner.qss`will cause the runner not to pick up on it? This is the kind of thing I want to document in that wiki.

How do the panel, runner, etc. know where to look for a stylesheet, or even _to_ look for one? I'm trying ripgrep on the source tree of lxqt-panel and finding no references to specific paths or `lxqt-panel.qss` anywhere.

**Reply by tsujan** [reply]
`liblxqt` defines `LXQt::Application` and applies to it the style sheet (if existing) whose name is that of the executable file. LXQt Panel and LXQt Runner are LXQt apps in their codes but, for example, PCManFM-Qt isn't. Therefore, PCManFM-Qt can't have a style sheet.

As far as I remember, LXQt Panel , LXQt Runner, lxqt-notificationd and lxqt-config are `LXQt::Application`, but there may be more, and other apps might be added to the list in future.
